File: homework3/hw3.m
pkg load statistics
load hw3_train.dat
load hw3_test.dat

% 13-15
% linear regression
N = 1000;
x = 2 * rand(N, 2) - 1;
noise = rand(N, 1);
noise(noise <= .1) = -1;
noise(noise > .1) = 1;
y = sign(x(:, 1).^2 + x(:, 2).^2 - .6) .* noise;
positive = find(y >= 0);
negive = find(y < 0);
plot(x(positive, 1), x(positive, 2), 'bo', x(negive, 1), x(negive, 2), 'r+');

X = [ones(N, 1) x];
w = pinv(X' * X) * X' * y;
yt = sign(X * w);
printf('Linear regress error is %f\n', sum(yt != y) / N);
pause;
printf('Plot linear fit on data\n');
f = @(w, x, y) w(1) + w(2) * x + w(3) * y;
plot_fit(f, w, -1:.01:1, -1:.01:1);
pause;

% feature transform
X = [ones(N, 1) x x(:, 1) .* x(:, 2) x.^2];
w = pinv(X' * X) * X' * y;
yt = sign(X * w);
printf('Quadratic feature transform error is %f\n', sum(yt != y) / N);
pause;
printf('Plot quadratic fit on data\n');
f = @(w, x, y) w(1) + w(2) * x + w(3) * y + w(4) * (x .* y) + w(5) * x.^2 + w(6) * y.^2;
plot_fit(f, w, -1:.01:1, -1:.01:1);
% hold on;
% [x1 x2] = meshgrid(-1:.01:1);
% z = w(1) + w(2) * x1 + w(3) * x2 + w(4) * (x1 .* x2) + w(5) * x1.^2 + w(6) * x2.^2;
% contour(x1, x2, z, [0 0], 'r-', 'LineWidth', 2);
% hold off;

% logistic regress
alpha = .01;
T = 2000;
x = hw3_train(:, 1:end - 1);
y = hw3_train(:, end);
N = size(y, 1);
X = [ones(N, 1) x];
w = zeros(size(X, 2), 1);
err_array = zeros(T, 1);
for i = 1:T
    w = w + alpha * -1 * get_deriv(w, X, y, mod(i, N) + 1); 
    [h p] = hyperthesis(w, X);
    err_array(i) = sum(h != y) / N;
end
plot(1:T, err_array, '+');
x = hw3_test(:, 1:end - 1);
y = hw3_test(:, end);
N = size(y, 1);
X = [ones(N, 1) x];
[h p] = hyperthesis(w, X);
err = sum(h != y) / N;
printf('Logistic regression error is %f\n', err);

File: homework3/hw3.R
source("common.R", echo = F)
source("homework3/get_deriv.R", echo = F)
source("homework3/sigmoid.R", echo = F)
source("homework3/predict.R", echo = F)

hw3_train <-  as.matrix(read.table("homework3/hw3_train.dat", stringsAsFactors = T))
hw3_test <-  as.matrix(read.table("homework3/hw3_test.dat", stringsAsFactors = T))

# 13-15 linear regression
N <- 1000
x <- matrix(2 * runif(2 * N) - 1, nrow = N)
noise <- as.matrix(runif(N))
noise[noise <= .1] <- -1
noise[noise > .1] <- 1
# y <- sign(v(x, 1) ^ 2 + v(x, 2) ^ 2 - .6) * noise
y <- (v(x, 1) ^ 2 + v(x, 2) ^ 2 - .6) * noise
yy <- sign(y)

X <- cbind(rep(1, N), x)
w <- solve(t(X) %*% X) %*% t(X) %*% y
yt <- sign(X %*% w)
message("Linear regress error is ", sum(yt != yy) / N)
pause()
message("Plot linear fit on data")
f <- function(x1, x2) w[1] + w[2] * x1 + w[3] * x2
plot_fit(f, x, yy)
pause()

# feature transform
X <- cbind(rep(1, N), x, v(x, 1) * v(x, 2), x ^ 2)
w <- solve(t(X) %*% X) %*% t(X) %*% y
yt <- sign(X %*% w)
message("Linear regression after feature transform error is ", sum(yt != yy) / N)
pause()
message("Plot feature transform line fit on data")
f <- function(x1, x2) w[1] + w[2] * x1 + w[3] * x2 + w[4] * x1 * x2 + w[5] * x1 ^2 + w[6] * x2 ^ 2
plot_fit(f, x, yy)
pause()

# logistic regression
message("Compute logistic train error")
alpha <- .01
T <- 2000
m <- nrow(hw3_train)
n <- ncol(hw3_train)
x <- hw3_train[, 1:n-1]
y <- v(hw3_train, n)
X <- cbind(rep(1, m), x)
w <- as.matrix(rep(0, n))
err_array <- rep(0, T)
for (i in 1:T) {
  cat("=")
  w <- w + alpha * -1 * get_deriv(w, X, y)
  yp <- predict(w, X)
  err_array[i] <- sum(yp != y) / m
}
plot(1:T, err_array, pch = 20)
pause()
message("Compute logistic test error")
x <- hw3_test[, 1:n-1]
y <- v(hw3_test, n)
m <- nrow(hw3_test)
X <- cbind(rep(1, m), x)
yp <- predict(w, X)
message("Logistic regression error is ", sum(yp != y) / m)